### PR TITLE
Refactor cycle detection traversal

### LINKED ISSUE #74: Refactor find_cycle function to reduce nested closure complexity
## Problem
The nested closures in `find_cycle` result in deep, complex control flow that is difficult to follow and maintain.

## Proposed Solution
Simplify the traversal logic by either:
- Moving the traversal into a separate type
- Breaking out smaller, focused functions
- Reducing nesting levels

This would ease comprehension and make the code more maintainable.

## Context
- Identified during code review in PR #69
- Comment: https://github.com/leynos/netsuke/pull/69#discussion_r13931
- Issue: https://github.com/leynos/netsuke/issues/74

### PR DESCRIPTION
**Pull request by leynos**
## Summary

Refactors IR cycle detection so traversal is driven through a reusable `CycleDetector::detect()` API with deterministic graph iteration, stronger state cleanup guarantees, and expanded test coverage.

## Functional Changes

- `analyse()` now delegates cycle discovery to `CycleDetector::detect()` instead of manually iterating target keys.
- Cycle detection now visits target nodes in sorted order, making the “first” reported cycle deterministic across runs regardless of `HashMap` iteration order.
- `CycleDetector::detect()` resets traversal state before each run, allowing repeated calls on the same detector to produce consistent results.
- DFS traversal now ensures the recursion stack is unwound even when a cycle is found, preventing stale stack state after early cycle detection.
- Cycle traversal now works with borrowed `Utf8Path` references internally while still storing owned paths in traversal state and reports.
- Missing dependencies are still collected as `(dependent, missing_dep)` pairs before the first detected cycle, and are now also emitted as debug-level tracing events.
- `canonicalize_cycle()` was simplified to remove the duplicate closing node before rotation, then re-close the cycle after rotating to the lexicographically smallest node.

## Tests

- Adds property-based tests for cycle canonicalization:
  - idempotency
  - rotation invariance
  - smallest-node-first ordering
  - closed-cycle output
- Adds property-based and targeted tests for detector behavior:
  - deterministic cycle detection
  - repeated `detect()` calls reset state
  - acyclic graphs return `None`
  - cyclic graphs return `Some`
  - stack is empty after both acyclic traversal and cycle detection
  - missing dependencies are reported before a later detected cycle
  - multiple disjoint cycles produce a valid canonical cycle

## Documentation

- Adds an IR cycle detection section to the developers guide describing:
  - `analyse()` as the entry point
  - `CycleDetectionReport`
  - `CycleDetector::new()` and `CycleDetector::detect()`
  - canonical cycle normalization
- Updates the design document to describe the new `detect` API.
- Normalizes several documentation spellings from British to US English, such as “serialisation” → “serialization” and “canonicalisation” → “canonicalization.”

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[src/ir.rs](https://github.com/leynos/netsuke/pull/163/files#diff-2c0aa19129c464ca53b35d941e7f401018ea4a56d2fc05815cca8c0920ae93fa)

Comment on lines +653 to +667

~~~rust
    fn canonicalize_cycle_handles_reverse_direction() {
        let cycle = vec![
            PathBuf::from("c"),
            PathBuf::from("b"),
            PathBuf::from("a"),
            PathBuf::from("c"),
        ];
        let canonical = canonicalize_cycle(cycle);
        let expected = vec![
            PathBuf::from("a"),
            PathBuf::from("c"),
            PathBuf::from("b"),
            PathBuf::from("a"),
        ];
        assert_eq!(canonical, expected);
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/69281/delta?repo-id=1160466&review-id=163&biomarker=Code+Duplication&filename=src%2Fir.rs&method=)
The module contains 2 functions with similar structure: tests.canonicalize_cycle_handles_reverse_direction,tests.canonicalize_cycle_rotates_smallest_node

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

~~~
Please address the comments from this code review:

## Individual Comments

### Comment 1
<location> `src/ir.rs:540-541` </location>
<code_context>
+            }
         }
+
+        self.stack.pop();
+        self.states.insert(node, VisitState::Visited);
+        None
     }
</code_context>

<issue_to_address>
**issue (bug_risk):** Potential issue if visit returns early: stack may not be properly unwound.

If visit exits early due to a cycle, self.stack.pop() is skipped, leaving the stack inconsistent. Ensure stack is always unwound, even on early returns.
</issue_to_address>

### Comment 2
<location> `src/ir.rs:562` </location>
<code_context>
         .enumerate()
         .min_by(|(_, a), (_, b)| a.cmp(b))
         .map_or(0, |(idx, _)| idx);
+    cycle.pop();
     cycle.rotate_left(start);
-    if let (Some(first), Some(slot)) = (cycle.first().cloned(), cycle.get_mut(len)) {
</code_context>

<issue_to_address>
**issue:** cycle.pop() may remove necessary information for single-node cycles.

This logic may fail for cycles of length one or two, such as self-loops. Please verify that all cycle sizes are handled correctly.
</issue_to_address>

### Comment 3
<location> `docs/netsuke-design.md:1086` </location>
<code_context>
+   deterministic error messages.
+
+   Traversal state is managed by a small `CycleDetector` helper struct. This
+   type owns the recursion stack and visitation map, allowing the traversal
+   functions to remain focused and easily testable.

</code_context>

<issue_to_address>
**suggestion (review_instructions):** This bullet paragraph exceeds 80 columns and should be wrapped for readability.

Please wrap this paragraph so that no line exceeds 80 columns, in accordance with the documentation formatting guidelines.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Paragraphs and bullets must be wrapped to 80 columns

</details>
</issue_to_address>
~~~

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (3 errors, 3 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Tests omit internal state verification when cycles detected. Test 2 checks stack.is\_empty() only for acyclic graphs. No test verifies stack/state consistency on cycle early returns. | Add test calling detect() with cycle and asserting stack.is\_empty() afterwards—would fail if cycle detection didn't clean stack properly. |
| Module-Level Documentation | ❌ Error | Module docstring in src/ir/cycle.rs lacks clear explanation of utility/function and does not describe relationships to other components (BuildEdge, graph module). | Expand module documentation to explain: the `analyse` function's role in detecting cycles in BuildEdge graphs, what `CycleDetectionReport` contains, and how `CycleDetector` manages traversal state. |
| Unit Architecture | ❌ Error | visit() early returns at line 93 without popping stack pushed at line 84 when cycle found in dependencies, violating the requirement that fallible operations leave state properly unwound. | Ensure stack.pop() is called on all exit paths by restructuring match/if-let logic or using guard patterns to guarantee cleanup before returning Some(cycle). |
| Docstring Coverage | ⚠️ Warning | Docstring coverage is 16.67% which is insufficient. The required threshold is 80.00%. | Write docstrings for the functions missing them to satisfy the coverage threshold. |
| Developer Documentation | ⚠️ Warning | CycleDetector refactoring documented in netsuke-design.md but missing from docs/developers-guide.md as required for internal API changes. | Add section to docs/developers-guide.md documenting ir::cycle module, CycleDetector type, and its detect() API for developers maintaining IR code. |
| Testing (Property / Proof) | ⚠️ Warning | PR introduces canonicalisation invariants without property-based testing. Proptest available but unused; only 2 hardcoded test cases fail to verify idempotence, equivalence, and edge cases. | Add proptest property tests to verify: (1) canonicalize idempotence, (2) all cycle rotations canonicalise identically, (3) edge cases (2–10 node cycles), and (4) lexicographical ordering invariant across generated cycle permutations. |

File: src/ir.rs [hunk 1]
```diff
@@ -581,33 +571,99 @@ fn canonicalize_cycle(mut cycle: Vec<PathBuf>) -> Vec<PathBuf> {
...
-            explicit_outputs: vec![PathBuf::from("b")],
-            implicit_outputs: Vec::new(),
-            order_only_deps: Vec::new(),
-            phony: false,
-            always: false,
-        };
-        targets.insert(PathBuf::from("a"), edge_a);
-        targets.insert(PathBuf::from("b"), edge_b);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/ir.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/ir.rs [hunk 2]
```diff
@@ -478,85 +478,74 @@ enum VisitState {
...
+                if let Some(cycle) = self.visit(dep.clone()) {
+                    return Some(cycle);
+                }
+            }
         }
+
+        self.stack.pop();
+        self.states.insert(node, VisitState::Visited);
+        None
     }
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/ir.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/ir.rs [hunk 3]
```diff
@@ -478,85 +478,74 @@ enum VisitState {
...
+                    .position(|n| n == &node)
+                    .expect("node should be present in the traversal stack");
+                let mut cycle: Vec<PathBuf> = self.stack.iter().skip(idx).cloned().collect();
+                cycle.push(node.clone());
+                return Some(canonicalize_cycle(cycle));
+            }
+            None => {
+                self.states.insert(node.clone(), VisitState::Visiting);
+            }
         }
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/ir.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/ir.rs [hunk 4]
```diff
@@ -478,85 +478,74 @@ enum VisitState {
...
                 continue;
             }
-
-            if let Some(cycle) = visit(targets, dep, stack, states) {
+            if let Some(cycle) = self.visit(node.clone()) {
                 return Some(cycle);
             }
         }
         None
     }
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/ir.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

docs/netsuke-design.md (1)

> `1-1715`: **Use en‑GB Oxford “-ize” spelling consistently in prose.**
> 
> The style guide specifies en‑GB‑oxendict with “-ize/-our”. Replace instances such as “serialise/deserialise/serialisation” with “serialize/deserialize/serialization” across the docs.

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### src/ir/cycle.rs (1)

`340-412`: _🛠️ Refactor suggestion_ | _🟠 Major_ | _🏗️ Heavy lift_
**Split this test-heavy module to satisfy the 400-line limit.**
Move `#[cfg(test)] mod tests` into a separate test module/file and keep

`src/ir/cycle.rs` under 400 lines.
As per coding guidelines: "Files must not exceed 400 lines in length".

##### 🤖 Detailed instructions

Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@src/ir/cycle.rs` around lines 340 - 412, The file exceeds 400 lines because
the large test module `mod property_tests` (which defines `node_names`,
`make_cycle` and multiple proptest cases exercising `canonicalize_cycle`) is in
`src/ir/cycle.rs`; move this entire `#[cfg(test)] mod property_tests` block into
a new test-only file (e.g., create a new module file such as
`src/ir/cycle_tests.rs` or `tests/cycle.rs`) and keep only `#[cfg(test)] mod
property_tests;` or a small re-export in `src/ir/cycle.rs`; ensure the relocated
file imports the same symbols (`use super::super::canonicalize_cycle; use
super::path;`) or adjusts paths so `node_names`, `make_cycle`, and all
`proptest!` tests compile and run unchanged.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (2 errors, 5 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Test cycle\_detector\_stack\_is\_empty\_after\_cycle\_detected is vacuous—would pass even if the early-return bug at line 117 existed, as outer scopes accidentally clean up the stack. | Add stronger test verifying stack.pop() aligns with stack.push() at each recursion level, not just final state, to catch the identified stack-unwinding bug. |
| Module-Level Documentation | ❌ Error | Module docstring for `src/ir/cycle.rs` is present but insufficient; PR objectives explicitly flag it as failing and require expansion via provided AI-agent prompts, which were not implemented. | Expand the module-level docstring in `src/ir/cycle.rs` to provide more comprehensive coverage of the module's design rationale, usage patterns, and integration points within the IR lowering pipeline, as outlined in the review feedback. |
| Linked Issues check | ⚠️ Warning | Critical unresolved issue: CycleDetector::visit() contains a stack-unwinding bug where early return on cycle detection skips self.stack.pop(), leaving the stack inconsistent. The PR description and code comments indicate awareness of this defect, yet the fix remains unimplemented in the current changeset. | Assign the find\_map result to a local variable, call self.stack.pop() unconditionally after the loop, and insert VisitState::Visited only when no cycle is found before returning. |
| Developer Documentation | ⚠️ Warning | CycleDetector refactoring lacks a developer-guide section; design doc updated, module doc present in cycle.rs, but developer's guide omits ir::cycle module documentation. | Add a section to docs/developers-guide.md documenting the ir::cycle module, CycleDetector struct, detect() API, and cycle canonicalisation strategy for maintainers. |
| Testing (Property / Proof) | ⚠️ Warning | CycleDetector.detect() introduces invariants (deterministic traversal, stack consistency) lacking property tests; only canonicalize\_cycle has proptest coverage. | Add proptest tests for CycleDetector.detect() covering arbitrary graphs, determinism across varied node orderings, and stack invariants. |
| Performance And Resource Use | ⚠️ Warning | PR introduces O(n log n) algorithmic regression by cloning and sorting all target keys. Custom check requires material regressions justified with complexity notes and benchmarks—none provided. | Add a code comment to `detect()` explaining the O(n log n) sorting trade-off for determinism, noting negligible impact for typical build graphs (100–10K targets), or provide a benchmark. |
| Architectural Complexity And Maintainability | ⚠️ Warning | CycleDetector appropriately encapsulates traversal state and reduces nesting. However, PR objectives document a critical unresolved bug: visit() skips stack cleanup when detecting cycles. | Assign the cycle-detection result to a local variable, unconditionally call stack.pop(), and return the cycle to ensure stack consistency. |

File: src/ir.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (leynos) [reply src/ir.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/ir.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (leynos) [reply src/ir.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/ir.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (leynos) [reply src/ir.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[src/ir/cycle_property_tests.rs](https://github.com/leynos/netsuke/pull/163/files/e95414010f8b58f51a10e9bc2c5b5f9033315272#diff-d737b2305d0d2283233f225dae0f5923851689474ae6d04ea36d6afcd3f0e6b4)

Comment on file

~~~rust
//! Property tests for IR cycle detection.
~~~

❌ New issue: [String Heavy Function Arguments](https://codescene.io/projects/69281/delta?repo-id=1160466&review-id=163&biomarker=String+Heavy+Function+Arguments&filename=src%2Fir%2Fcycle_property_tests.rs&method=)
In this module, 60.0% of all arguments to its 17 functions are strings. The threshold for string arguments is 39.0%

File: src/ir/cycle_property_tests.rs [hunk 5]
```diff
@@ -0,0 +1,187 @@
+use proptest::prelude::*;
+use std::collections::HashMap;
+
+use super::super::{canonicalize_cycle, CycleDetector};
+use super::{build_edge, path};
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/ir/cycle_property_tests.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/ir/cycle_property_tests.rs [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    ));
+    cycle
+}
+
+fn check_canonicalize_cycle(input: &[&str], expected: &[&str]) {
+    let cycle: Vec<camino::Utf8PathBuf> = input.iter().map(|&s| path(s)).collect();
+    let canonical = canonicalize_cycle(cycle);
+    let want: Vec<camino::Utf8PathBuf> = expected.iter().map(|&s| path(s)).collect();
+    assert_eq!(canonical, want);
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/ir/cycle_property_tests.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/ir/cycle_property_tests.rs [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+#[test]
+fn find_cycle_detects_one_of_multiple_disjoint_cycles() {
+    let mut targets = HashMap::new();
+    targets.insert(path("p"), build_edge(&["q"], &[], "p"));
+    targets.insert(path("q"), build_edge(&["p"], &[], "q"));
+    targets.insert(path("x"), build_edge(&["y"], &[], "x"));
+    targets.insert(path("y"), build_edge(&["x"], &[], "y"));
+
+    assert!(CycleDetector::find_cycle(&targets).is_some());
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/ir/cycle_property_tests.rs:212]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/ir/cycle_property_tests.rs [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+
+#[test]
+fn find_cycle_detects_one_of_multiple_disjoint_cycles() {
+    let mut targets = HashMap::new();
+    targets.insert(path("p"), build_edge(&["q"], &[], "p"));
+    targets.insert(path("q"), build_edge(&["p"], &[], "q"));
+    targets.insert(path("x"), build_edge(&["y"], &[], "x"));
+    targets.insert(path("y"), build_edge(&["x"], &[], "y"));
+
+    assert!(CycleDetector::find_cycle(&targets).is_some());
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/ir/cycle_property_tests.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@kody start-review

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (3 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Unit And Behavioural) | ⚠️ Warning | Unit tests drive internal CycleDetector implementation via test-only wrapper find\_cycle rather than testing the public analyse() API boundary directly. | Add unit tests directly exercising analyse() to verify CycleDetectionReport fields, rather than testing private CycleDetector traversal state. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | PR lacks snapshot tests for CircularDependency error message output. While property and unit tests cover logic, formatted error messages lack snapshot verification. | Add insta snapshot tests for CircularDependency error formatting and diagnostic JSON output to verify user-facing error presentation. |
| Architectural Complexity And Maintainability | ⚠️ Warning | CycleDetector is a private, single-use abstraction without demonstrated reuse. Net complexity increased (742 lines) without simplifying the call site or clearly justifying the abstraction. | Flatten CycleDetector into analyse() or document the near-term reuse path and architecture rationale that justifies introducing this stateful detector abstraction now. |
